src/socket.rs

```diff
@@ -0,0 +1,335 @@
+use crate::arch::Syscalls;
+use crate::close;
+use crate::{result, syscall};
+use core::mem::size_of_val;
+use libc::{
+    sockaddr, socklen_t, AF_INET, AF_INET6, SOCK_CLOEXEC, SOCK_DGRAM, SOCK_NONBLOCK, SOCK_RAW,
+    SOCK_RDM, SOCK_SEQPACKET, SOCK_STREAM,
+};
+use std::net::SocketAddr;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::{
+    io,
+    mem::{self, MaybeUninit},
+    ptr,
+};
+
+pub struct Fd(usize);
+
+impl AsRawFd for Fd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.0 as RawFd
+    }
+}
+
+impl AsRawFd for &Fd {
+    fn as_raw_fd(&self) -> RawFd {
+        (*self).as_raw_fd()
+    }
+}
+
+impl Drop for Fd {
+    fn drop(&mut self) {
+        unsafe {
+            close(self).ok();
+        }
+    }
+}
+
+fn socket_addr(addr: SocketAddr) -> (*const sockaddr, socklen_t) {
+    match addr {
+        SocketAddr::V4(ref addr) => (
+            addr as *const _ as *const sockaddr,
+            size_of_val(addr) as socklen_t,
+        ),
+        SocketAddr::V6(ref addr) => (
+            addr as *const _ as *const sockaddr,
+            size_of_val(addr) as socklen_t,
+        ),
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+#[repr(i32)]
```

Comment:
I prefer to cast to `isize` instead of using `repr` because repr make the exact size+alignment a requirement.
but we don't need it to be a requirement. we just want the right flag when *casting* to the right type.

so casting to `isize` and not specifying exact repr let's the compiler optimize it nicely: https://play.rust-lang.org/?gist=0afefa655c772e7037f5dc8d0fce9a19